_posts/2017/2017-11-26-choose-uuids-for-model-ids-in-rails.md

```diff
@@ -0,0 +1,83 @@
+---
+title: 'Choose UUIDs for model ids in Rails'
+description: 'A little more complexity, a little more security'
+layout: article
+category: ruby
+image:
+  base: '2017/choose-uuids-for-model-ids-in-rails'
+  alt: 'Rainbow UUIDs'
+---
+
+Ruby on Rails has had the [ability](https://github.com/rails/rails/pull/21762) to use UUIDs ([Universally Unique IDentifiers](https://en.wikipedia.org/wiki/Universally_unique_identifier)) for ActiveRecord models since version 5.0.
+
+These are a native column type in PostgreSQL, there's more details on native Postgres types in the [Rails Guides](http://guides.rubyonrails.org/active_record_postgresql.html).
+
+## Instead of…
+
+…using Rails default incrementing integer IDs.
+
+
+## Use…
+
+…PostgreSQL’s UUID support.
+
+
+### Enable extension
+
+`bin/rails g migration enable_extension_for_uuid`
+
+```ruby
+class EnableExtensionForUuid < ActiveRecord::Migration[5.1]
+  def change
+    enable_extension 'pgcrypto' unless extension_enabled?('pgcrypto')
+  end
+end
+```
+
+Change the default column type for primary keys
+
+Configure your migration generator to set id: :uuid for new tables:
+
+
+### `config/initializers/generators.rb`
+
+```ruby
+Rails.application.config.generators do |g|
+  g.orm :active_record, primary_key_type: :uuid
+end
+```
+
+
+### In future migrations
+
+You'll have to use `type: :uuid` when creating relations.
+
+```ruby
+class AddNewTable < ActiveRecord::Migration[5.1]
+  def change
+    create_table :related_model do |t|
+      t.references :other, type: :uuid, index: true
+    end
+  end
+end
+```
+
+
+## But why?
+
+This is a case where you are making a choice toward a little more complexity, but for good reasons.
```

Comment:
Put this statement at the end.

By this point, I've read so much setup..and I'm finally at 'But why?' and we're still not there yet.

In fact, whilst reading this post, I've been thinking that maybe you should restructure these posts. The 'how' should come after the 'but why?' because you need to convince the reader first that it's something worth doing..then show them how to do it.

I've been reading the setup and being like: yeah sure, but why would I want to do this? Because it's you I'm gonna read anyway, and I'm almost looking forward to the 'but why?' but for other people they may lose focus and stop reading.